flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -89,6 +91,25 @@
 		super(index, parent);
 	}
 
+	@Override
+	public void initializeState() throws IOException, InterruptedException {
+		ReadResult readResult = ReadResult.HAS_MORE_DATA;
+		while (readResult == ReadResult.HAS_MORE_DATA) {
+			BufferBuilder bufferBuilder = parent.getBufferPool().requestBufferBuilderBlocking();
+			BufferConsumer bufferConsumer = bufferBuilder.createBufferConsumer();
+			readResult = parent.getChannelStateReader().readOutputData(subpartitionInfo, bufferBuilder);
```

Comment:
I think it is not so necessary to decouple `ResultPartition` with `ResultSubpartition`, because the `ResultSubpartition` is derived from `ResultPartition` and it can get everything from parent by design. In the constructor of `ResultSubpartition`, `ResultPartition` is already passed as argument and defined as protected variable to be referenced directly by any kind of subpartitions. So we do not destroy the existing rules or add any new rules. Also for tests, the separate `PipelinedSubpartitionTest` still needs to construct parent `ResultPartition` in practice.

But your comment makes me think of another possible option, to pass `ChannelStateReader` into `ResultPartitionWriter#initializeState`, then we do not need to pass it into constructor to touch many components.

Reply:
> But your comment makes me think of another possible option, to pass ChannelStateReader into ResultPartitionWriter#initializeState, then we do not need to pass it into constructor to touch many components.

This is exactly what I meant by 
> Why not to pass ChannelStateReader as a parameter to initializeState then?

So, instead of having `channelStateReader` field anywhere I'd pass it from `StreamTask` 
to `ResultPartitionWriter.initializeState()` 
to `ResultPartition.initializeState()` 
to `ResultSubpartition.initializeState()` 

Sorry for the confusion